### ISSUE TITLE
Possibly simplify development process

### ISSUE THREAD

**Issue by jordansissel** [elastic/logstash#4746]
Problem as I see it: `rake bootstrap` and then `rake test:install-core` take a looooooong time. I wish it to be faster.

History:

Way back when, we decided to continue supporting MRI as a runtime Ruby for Logstash. We have since reversed that decision and gone much deeper into JRuby/Java and we no longer even attempt MRI compatibility.

Part of the compatibility work for MRI+JRuby I did for [The Great Rake Refactor](920a9e93621495d3485b087d81ead9c8ebe656be) was to make it a bit easier to switch to different rubies and also to help make bootstrapping for non-jruby-users easier.

I'm not sure we need much of this anymore.

----

My hope is that we can improve the bootstrapping process and maybe even boil it down to a `bundle install` -- maybe? That, or make the plugin installation faster.

Thoughts?

**Comment by purbon** [comment]
Moin moin,
   I see some slowness paint points here,

* Need to bootstrap the env with an embed jruby version, mostly download time. 
* Install bundler
* run bundler install command.
* Compile the current config.
* Install core plugins for test.

Download the jruby version is something we have to do somehow, and will never become as much faster as your internet connection. My reasonable guess is that this, together with ```install-core``` are the major offenders in terms of time. _need prove by getting numbers_.

One thing we have to explore is to avoid the need of using real plugins for test, this is nice for some test, but for most of ours is not much required as our plugin usage is not that elaborate. We're also start to having a mix of "dummy" plugins vs real ones (see https://github.com/elastic/logstash/blob/master/logstash-core/spec/logstash/pipeline_spec.rb#L6-L103 ). This will bring some speed, we could also include this "required" plugins inside the Gemfile and leverage only one bundle install call.

what do you think?

**Comment by ebuildy** [comment]
What about a Docker image ?

**Comment by jordansissel** [comment]
@ebuildy I don't use docker, so that won't help me much. Plus it feels doesn't really help the bootstrapping problem (unless we publish a docker image for every commit on each branch?). That also means it'll require another moving part (docker) to do some logstash dev work, and I'm hoping to reduce parts.

**Comment by jordansissel** [comment]
@purbon ++ to not needing real plugins for most core tests.

**Comment by purbon** [comment]
@jordansissel agree, on the other side, not sure how to speed up things (as we bundle a jruby version with LS) .... this process is bound to internet connection. we could however investigate bundler speed up also?

**Comment by colinsurprenant** [comment]
We are mixing two issues here: 1- the process **complexity** and 2- the process **speed**. I agree both directly affect the user/developer experience.

I agree we should revisit the bootstrapping process **complexity** from a developer "getting started" point of view. I also believe it can be simplified.

I also think we should ditch MRI support. JRuby is now required so there is no point for supporting MRI just for the case where your default interpreter is set to MRI. We should just fail when the current interpreter is not JRuby, this would simplify the bootstrap **code** anyway.

For the process speed, there is the initial case, where you bootstrap from a *virgin* environment and there is the on-going development process.

1. Virgin environment bootstrap

  Basically 2 things are somewhat slow: a) the JRuby download and b) the plugins installation. Typically, unless you continuously wipe your environment, these two steps should happen basically only once. I think it is OK to pay an initial bootstrapping cost to properly setup the environment. I think the problem is more with understanding (and **documenting**) when and under what conditions an environment should be reset. When you don't reset, the flow is somewhat OK IMO.

  For the JRuby download, this should **really** only happen once. For the plugins, there is no magic here if we need to install the **latest** of **all plugins** from a virgin environment - we can't accelerate this. There are 2 things that can improve this: 1- correctly understanding (and **documenting**) where a full plugins install is required and 2- see #4787 - we could potentially provide plugins installation packs which **do not** require going through a **bundle install** process but instead just copy the `vendor/bundle`, `Gemfile` and `Gemfile.jruby-1.9.lock` in place.  

2. On-going development process

  As explained in (1) under normal conditions, if the environment is not wiped, updating a plugin reference to point to a local dev directory or running specs is OK I believe - maybe missing documentation or guidelines? 

Obviously, our environment has some intrinsic complexity, I believe we can revisit and potentially improve the process and associated tasks, documentation, etc, but at the same time there will always be a minimum involvement expected from a developer.
